packages/react-storage/src/components/StorageBrowser/views/hooks/useSearch.ts

```diff
@@ -0,0 +1,77 @@
+import React from 'react';
+import debounce from 'lodash/debounce';
+
+const DEFAULT_DELAY = 300;
+export interface InitialSearchValues<T> {
+  searchKey: keyof T;
+  searchTerm?: string;
+  debounceDelay?: number;
+}
+
+interface UseSearchProps<T> {
+  initialValues: InitialSearchValues<T>;
+  items: T[];
+  onSearch: (term: string, includeSubfolders?: boolean) => void;
+}
+
+interface UseSearchResult<T> {
+  filteredItems: T[];
+  searchTerm: string;
+  onSearch: (term: string, includeSubfolders?: boolean) => void;
+}
+
+/**
+ * Custom hook to search a list of items based on a search key and search term.
+ * @template T - The type of items in the search list.
+ * @param {UseSearchProps<T>} props
+ * @returns {UseSearch<T>} - The search results, current term, and handler.
+ */
+export function useSearch<T>({
+  items,
+  initialValues: initialValuesProp,
+  onSearch,
+}: UseSearchProps<T>): UseSearchResult<T> {
+  const [initialValues] = React.useState(initialValuesProp);
+  const {
+    searchTerm: initialSearchTerm = '',
+    searchKey,
+    debounceDelay = DEFAULT_DELAY,
+  } = initialValues;
+
+  const [searchTerm, setSearchTerm] = React.useState(initialSearchTerm);
+
+  const filteredItems = React.useMemo(
+    () =>
+      items.filter((item) => {
+        const test = item[searchKey];
+        if (typeof test === 'string') {
+          return test.includes(searchTerm);
+        }
+      }),
+    [items, searchKey, searchTerm]
+  );
+
+  const debouncedHandler = React.useMemo(
+    () =>
+      debounce((term: string, includeSubfolders?: boolean) => {
```

Comment:
Know we chatted offline but still not quite sure we need to add debouncing here. Our default implementation only runs `onSearch` on button click, and the button should be disabled while a search is in progress

Reply:
I'll remove it for now, just to simplify this given our current UI. It's easy enough to add later.